broker/httpd/instrumentedWriter.go

```diff
@@ -0,0 +1,232 @@
+	RequestHeader                             http.Header
+	CustomRecords                             map[string]string
+}
+
+type LoggingWriter struct {
+	http.ResponseWriter
+	logRecord LogRecord
+}
+
+var (
+	rpcDurations = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name:       "service_http_request_processing_duration_seconds",
+			Help:       "RPC latency distributions.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{"code"},
+	)
+	rpcDurationsHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "service_http_request_response_size",
+			Help: "RPC latency distributions.",
+			//Buckets: prometheus.LinearBuckets(*normMean-5**normDomain, .5**normDomain, 20),
+			Buckets: prometheus.ExponentialBuckets(10.0, 1.5, 20),
+		},
+		[]string{"code"},
+	)
+)
+
+func init() {
+	// Register the summary and the histogram with Prometheus's default registry.
+	prometheus.MustRegister(rpcDurations)
+	prometheus.MustRegister(rpcDurationsHistogram)
+}
+
+func (r *LoggingWriter) Write(p []byte) (int, error) {
+	if r.logRecord.Status == 0 {
+		// The status will be StatusOK if WriteHeader has not been called yet
+		r.logRecord.Status = http.StatusOK
+	}
+	written, err := r.ResponseWriter.Write(p)
+	r.logRecord.Size += int64(written)
+	return written, err
+}
+
+func (r *LoggingWriter) WriteHeader(status int) {
+	r.logRecord.Status = status
+	r.ResponseWriter.WriteHeader(status)
+}
+
+// w.(accesslogger.LoggingWriter).SetCustomLogRecord("X-User-Id", "3")
+func (r *LoggingWriter) SetCustomLogRecord(key, value string) {
+	if r.logRecord.CustomRecords == nil {
+		r.logRecord.CustomRecords = map[string]string{}
+	}
+	r.logRecord.CustomRecords[key] = value
+}
+
+//  w.(accesslogger.LoggingWriter).SetUsername("alice")
+func (r *LoggingWriter) SetUsername(username string) {
+	r.logRecord.Username = username
+}
+
+// http.CloseNotifier interface
+func (r *LoggingWriter) CloseNotify() <-chan bool {
+	if w, ok := r.ResponseWriter.(http.CloseNotifier); ok {
+		return w.CloseNotify()
+	}
+	return make(chan bool)
+}
+
+func (r *LoggingWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	if hijacker, ok := r.ResponseWriter.(http.Hijacker); ok {
+		return hijacker.Hijack()
+	}
+	return nil, nil, fmt.Errorf("ResponseWriter doesn't support Hijacker interface")
+}
+
+// http.Flusher
+func (r *LoggingWriter) Flush() {
+	flusher, ok := r.ResponseWriter.(http.Flusher)
+	if ok {
+		flusher.Flush()
+	}
+}
+
+// http.Pusher
+func (r *LoggingWriter) Push(target string, opts *http.PushOptions) error {
+	pusher, ok := r.ResponseWriter.(http.Pusher)
+	if ok {
+		return pusher.Push(target, opts)
+	}
+	return fmt.Errorf("ResponseWriter doesn't support Pusher interface")
+}
+
+type Logger interface {
+	Log(record LogRecord)
+}
+
+type LoggingHandler struct {
+	handler   http.Handler
+	logger    Logger
+	logBefore bool
+}
+
+func NewLoggingHandler(handler http.Handler, logger Logger) http.Handler {
```

Comment:
Do you anticipate this and other functions being called from other packages? If not, please make them private.

Reply:
It think I do, there are a lot of other interfaces that this type should implement so that it can behave transparently